lib/Markup/Parser/MarkupLexical.swift

```diff
@@ -0,0 +1,113 @@
+import Foundation
+
+struct LexicalAnalyzer {
+    private let content: String
+    private var currentIndex: Int
+
+    init(content: String) {
+        self.content = content
+        self.currentIndex = 0
+    }
+
+    mutating func nextToken() -> (token: Token, line: String)? {
```

Comment:
Is it possible to make this side-effect free? It is mutating a state.

Reply:
In the end, we are using it as a scanner for content (like MarkupSyntax) - if we want to do that, we should change approach, to scan the next token given a content + position, or return a new struct with the new state.

Both options, I feel add extra complex to code, so I decided to keep this version saving only the current position for lexical analyzer (making mutating syntax analyzer too - it is using it).

Now, I finished the analyzer, I gonna go to review it again and see if it is possible without mess up the code :)